skopt/gp_opt.py

```diff
@@ -227,28 +231,48 @@ def gp_minimize(func, dimensions, base_estimator=None, acq="EI", xi=0.01,
     models = []
     n_model_iter = n_calls - n_total_init_calls
     for i in range(n_model_iter):
+        Xi_transform = space.transform(Xi)
         gp = clone(base_estimator)
 
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
-            gp.fit(space.transform(Xi), yi)
+            gp.fit(Xi_transform, yi)
 
         models.append(gp)
 
         if search == "sampling":
+
             X = space.transform(space.rvs(n_samples=n_points,
                                           random_state=rng))
+
+            if is_discrete_space:
+                # Choose the point that gives the best acquisition
+                # value AND has not been reevaluated at before.
+                # This prevents costly function reevaluations.
+                # while exploring more of the search space.
+                X_in_Xi_transform = in2d(X, Xi_transform)
+
+                # Highly unlikely corner case.
```

Comment:
Shall we really stop? if there are remaining calls, we might still find an unvisited point.